source/speech/__init__.py

```diff
@@ -1282,51 +1386,89 @@ def getControlFieldSpeech(attrs,ancestorAttrs,fieldType,formatConfig=None,extraD
 			getProps['columnHeaderText'] = attrs.get("table-columnheadertext")
-		return (getSpeechTextForProperties(_tableID=tableID, **getProps)
-			+ (" %s" % stateText if stateText else "")
-			+ (" %s" % ariaCurrentText if ariaCurrent else ""))
+		tableCellsequence = getPropertiesSpeech(_tableID=tableID, **getProps)
+		tableCellsequence.extend(stateTextSequence)
+		tableCellsequence.extend(ariaCurrentSequence)
+		return tableCellsequence
 
 	# General cases.
 	if (
-		(speakEntry and ((speakContentFirst and fieldType in ("end_relative","end_inControlFieldStack")) or (not speakContentFirst and fieldType in ("start_addedToControlFieldStack","start_relative"))))
-		or (speakWithinForLine and not speakContentFirst and not extraDetail and fieldType=="start_inControlFieldStack")
+		speakEntry and (
+			speakContentFirst
+			and fieldType in ("end_relative", "end_inControlFieldStack")
+		) or (
+				not speakContentFirst
+				and fieldType in ("start_addedToControlFieldStack", "start_relative")
+		)
+	) or (
+			speakWithinForLine
+			and not speakContentFirst
+			and not extraDetail
+			and fieldType == "start_inControlFieldStack"
 	):
 		out = []
 		content = attrs.get("content")
 		if content and speakContentFirst:
 			out.append(content)
 		if placeholderValue:
-			if valueText:
-				log.error("valueText exists when expected none: valueText:'%s' placeholderText:'%s'"%(valueText,placeholderText))
-			valueText = placeholderText
-		out.extend(x for x in (nameText,(stateText if speakStatesFirst else roleText),(roleText if speakStatesFirst else stateText),containerContainsText,ariaCurrentText,valueText,descriptionText,levelText,keyboardShortcutText) if x)
+			if valueSequence:
+				log.error(
+					f"valueSequence exists when expected none: "
+					f"valueSequence: {valueSequence!r} placeholderSequence: {placeholderSequence!r}"
+				)
+			valueSequence = placeholderSequence
+		out.extend(nameSequence)
+		out.extend(stateTextSequence if speakStatesFirst else roleTextSequence)
+		out.extend(roleTextSequence if speakStatesFirst else stateTextSequence)
+		out.append(containerContainsText)
+		out.extend(ariaCurrentSequence)
+		out.extend(valueSequence)
+		out.extend(descriptionSequence)
+		out.extend(levelSequence)
+		out.extend(keyboardShortcutSequence)
 		if content and not speakContentFirst:
 			out.append(content)
-		return CHUNK_SEPARATOR.join(out)
-		
-	elif fieldType in ("end_removedFromControlFieldStack","end_relative") and roleText and ((not extraDetail and speakExitForLine) or (extraDetail and speakExitForOther)):
+		return out
+
+	elif fieldType in (
+			"end_removedFromControlFieldStack", "end_relative"
+	) and roleTextSequence and (
+		(not extraDetail and speakExitForLine) or (extraDetail and speakExitForOther)
+	):
+		return [
 		# Translators: Indicates end of something (example output: at the end of a list, speaks out of list).
-		return _("out of %s")%roleText
+			_("out of %s") % " ".join(roleTextSequence)

```

Comment:
This will break if roleTextSequence contains speech commands.

Reply:
I'm not really sure what to do here. We could prepend the string "out of " and extend with roleTextSequence, but the "out of " string really is meaningless without subsequent text. Is there a way to do this, I'm perhaps jumping to conclusions, but I wonder if speech sequences should actually be "tree" like.  I'm probably missing something, @jcsteh would you be able to clarify for me?

Reply:
With tree like, do you mean having a list of strings/commands in another list that could either contain more lists or strings or commands?

Reply:
> With tree like, do you mean having a list of strings/commands in another list that could either contain more lists or strings or commands?

Yes, essentially. So that you can bundle groups of related speech text or commands (due to them having the same origin / context).  Perhaps we could do this with commands, perhaps this is the intention behind the `endUtteranceCommand`

Reply:
The "out of" indicator just doesn't work for roles which aren't described textually. For example, if we use a click sound for a link, "out of <click>" is just ridiculous. If the role doesn't produce text, we should just skip this altogether.

One question is what to do if the role has both text and speech commands. One could argue we should take both, but I don't think this is a good idea. For now, I think we should just only do this "out of" thing for roles that produce only text. In future, we may be able to come up with other options; e.g. if a role produces sound, we could have an "out of" sound so we end up with "<out of sound> <role sound>". I think that's getting way ahead of ourselves here though.